docs/engineering/Testing/ui-tests.md

```diff
@@ -0,0 +1,550 @@
+
+There is a general rule that we want to test the functional results of our code, not the implementation details, so that when things get refactored, our tests don't break unless we've actually done something that would effect how the site works for users. That means, as much as possible, when deciding what to test, we want to think about how our components work from the users' perspective and test the things that they see on screen/hear with a screen reader, interact with, or expect to work a certain way, rather than the internal logic of our components. If someone would file a bug report about it if it broke, you should test it. But this is also not the way to test layout or styling things like colors or sizes, so don't get too bogged down in details, even if things like colour are being used to convey some amount of meaning. For instance, if we are testing the Side Menu, we want to test that boards with new updates are marked correctly, and within that test we want to test both that the board links are accessibly labeled in a way that flags the presence of new updates for screen readers, and that the notification dot is visible for sighted users, but not the color of the notification dot in comparison to the darker color used for outdated updates.
+
+For a full example, here the descriptions of the tests for the BoardMenuSection component of the Side Menu:
+
+```typescript title="tests/20-SideMenu/BoardMenuSection.test.tsx"
+describe("Regular", () => {
+  test("Renders section with board menu items", async () => {
+    render(<Regular />);
+    // Test details
+  });
+
+  test("Board menu items link to boards", async () => {
+    render(<Regular />);
+    // Test details
+  });
+
+  test("Correctly renders board without updates", async () => {
+    render(<Regular />);
+    // Test details
+  });
+
+  test("Correctly marks boards with updates", async () => {
+    render(<Regular />);
+    // Test details
+  });
+
+  test("Correctly marks outdated boards with updates", async () => {
+    render(<Regular />);
+    // Test details
+  });
+
+  test("Correctly marks current board", async () => {
+    render(<Regular />);
+    // Test details
+  });
+
+  test("Correctly marks muted board", async () => {
+    render(<Regular />);
+    // Test details
+  });
+});
+
+describe("Empty", () => {
+  test("Renders empty section", async () => {
+    render(<Empty />);
+    // Test details
+  });
+});
+
+describe("Loading", () => {
+  test("Renders loading section", async () => {
+    render(<Loading />);
+    // Test details
+  });
+});
+```
+
+Similar tests are run on the Pinned Menu, and then the following tests are run on the whole Side Menu:
+
+```typescript title="tests/20-SideMenu/SideMenu.test.tsx"
```

Comment:
These are too long examples, and I don't think they're adding much. If you want to give examples, I'd simply point out what you might test as a list, without code.

You might say something like:
For example, for a board menu section (link to storybook) we might test:
- That the regular section is displayed
- That a board with updates is displayed
- That the current board is displayed

We should also tests that this is displaying correctly in context, without repeating the bulk of tests. For example, in Side Menu, we can write the following tests:
- Sections are displayed correctly
- Pinned menu is not rendered when turned off
- ...

(the language above sucks, but that's the idea)